ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/PeerDiscoveryAgent.java

```diff
@@ -53,282 +51,201 @@
+  private final PeerRequirement peerRequirement;
+  private final PeerBlacklist peerBlacklist;
+  private final NodeWhitelistController nodeWhitelistController;
   /* The peer controller, which takes care of the state machine of peers. */
-  private final PeerDiscoveryController controller;
+  protected Optional<PeerDiscoveryController> controller = Optional.empty();
+
   /* The keypair used to sign messages. */
-  private final SECP256K1.KeyPair keyPair;
+  protected final SECP256K1.KeyPair keyPair;
+  private final BytesValue id;
   private final PeerTable peerTable;
-  private final DiscoveryConfiguration config;
+  protected final DiscoveryConfiguration config;
 
   /* This is the {@link tech.pegasys.pantheon.ethereum.p2p.Peer} object holding who we are. */
   private DiscoveryPeer advertisedPeer;
-  /* The vert.x UDP socket. */
-  private DatagramSocket socket;
+  private InetSocketAddress localAddress;
 
   /* Is discovery enabled? */
   private boolean isActive = false;
+  private final Subscribers<Consumer<PeerBondedEvent>> peerBondedObservers = new Subscribers<>();
 
   public PeerDiscoveryAgent(
-      final Vertx vertx,
       final SECP256K1.KeyPair keyPair,
       final DiscoveryConfiguration config,
       final PeerRequirement peerRequirement,
       final PeerBlacklist peerBlacklist,
       final NodeWhitelistController nodeWhitelistController) {
-    checkArgument(vertx != null, "vertx instance cannot be null");
     checkArgument(keyPair != null, "keypair cannot be null");
     checkArgument(config != null, "provided configuration cannot be null");
 
     validateConfiguration(config);
 
-    final List<DiscoveryPeer> bootstrapPeers =
+    this.peerRequirement = peerRequirement;
+    this.peerBlacklist = peerBlacklist;
+    this.nodeWhitelistController = nodeWhitelistController;
+    this.bootstrapPeers =
         config.getBootstrapPeers().stream().map(DiscoveryPeer::new).collect(Collectors.toList());
 
-    this.vertx = vertx;
     this.config = config;
     this.keyPair = keyPair;
     this.peerTable = new PeerTable(keyPair.getPublicKey().getEncodedBytes(), 16);
-    this.controller =
-        new PeerDiscoveryController(
-            vertx,
-            this,
-            peerTable,
-            bootstrapPeers,
-            PEER_REFRESH_INTERVAL_MS,
-            peerRequirement,
-            peerBlacklist,
-            nodeWhitelistController);
+
+    id = keyPair.getPublicKey().getEncodedBytes();
   }
 
-  public CompletableFuture<?> start(final int tcpPort) {
-    final CompletableFuture<?> completion = new CompletableFuture<>();
+  protected abstract TimerUtil createTimer();
+
+  protected abstract CompletableFuture<InetSocketAddress> listenForConnections();
+
+  protected abstract CompletableFuture<Void> sendOutgoingPacket(
+      final DiscoveryPeer peer, final Packet packet);
+
+  public abstract CompletableFuture<?> stop();
+
+  public CompletableFuture<?> start() {
```

Comment:
wow- interesting, so this only uses netty now?

Reply:
No, but I moved all the vertx code into a subclass. Shouldn't be too much work to migrate to netty now.